File: 905.Sort-Array-By-Parity.py
# https://leetcode.com/problems/sort-array-by-parity/description/
#
# algorithms
# Easy (73.57%)
# Total Accepted:    10.3k
# Total Submissions: 14k
# beats 72.87% of python submissions


class Solution(object):
    def sortArrayByParity(self, A):
        """
        :type A: List[int]
        :rtype: List[int]
        """
        length = len(A)

        if length <= 1:
            return A

        head, tail = 0, length - 1

        while head < tail:
            while head < tail and A[head] & 1 == 0:
                head += 1
            while tail > head and A[tail] & 1 == 1:
                tail -= 1
            A[head], A[tail] = A[tail], A[head]
            head += 1
            tail -= 1

        return A

File: 905.Sort-Array-By-Parity.java
// https://leetcode.com/problems/sort-array-by-parity/
//
// algorithms
// Easy (73.23%)
// Total Accepted: 145,743
// Total Submissions: 199,020


class Solution {
    public int[] sortArrayByParity(int[] A) {
        int len = A.length;
        int preIdx = 0;

        for (int i = 0; i < len; i++) {
            if ((A[i] & 1) == 0) {
                swap(A, preIdx++, i);
            }
        }

        return A;
    }

    private void swap(int[] A, int i, int j) {
        int tmp = A[i];
        A[i] = A[j];
        A[j] = tmp;
    }
}